### ISSUE TITLE
[engine] NullPointerException thrown when using URLForwarding

### ISSUE THREAD

**Issue by philipp-brock-thalia** [craftercms/craftercms#4464]
When configuring some urls to be forwarded to others than the SiteContext.clear()-Method is throwing a NullPointerException.

### What we have done
We have configured a rewrite rule with tuckey inside the `/config/engine/urlrewrite.xml` like this:

    <urlrewrite>
        <rule>
            <from>^/test$</from>
            <to type="forward">/test-new</to>
        </rule>
    </urlrewrite>

### What happens
After configuring this rewrite the SiteContext.clear()-Method throws an NullPointerException when calling the "from" url because there is no context in threadLocal, when it should be cleared. See the following StackTrace

    14-Jan-2021 14:20:29.125 SCHWERWIEGEND [http-nio-8080-exec-9] org.apache.catalina.core.StandardWrapperValve.invoke Servlet.service() for servlet [Spring MVC Dispatcher Servlet] in context with path [] threw exception
    	java.lang.NullPointerException
    		at org.craftercms.engine.service.context.SiteContext.clear(SiteContext.java:162)
    		at org.craftercms.engine.servlet.filter.SiteContextResolvingFilter.doFilter(SiteContextResolvingFilter.java:88)
    		at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:347)
    		at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:263)
    		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
    		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
    		at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101)
    		at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:347)
    		at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:263)
    		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
    		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
    		at org.craftercms.commons.http.RequestContextBindingFilter.doFilter(RequestContextBindingFilter.java:78)
    		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
    		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
    		at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197)
    		at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
    		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
    		at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101)
    		at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:347)
    		at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:263)
    		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
    		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
    		at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101)
    		at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:347)
    		at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:263)
    		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
    		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
    		at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71)
    		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
    		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
    		at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:199)
    		at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
    		at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:543)
    		at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
    		at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
    		at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:688)
    		at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
    		at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
    		at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:615)
    		at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
    		at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:818)
    		at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1623)
    		at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
    		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    		at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
    		at java.lang.Thread.run(Thread.java:748)

The forwarding is done nevertheless, but the thrown error is not good because our logs are wasted by this and we are also not sure if everyhing works fine afterwards.

### Some additional information

- We're currently upgrading from CrafterCMS 3.1.4 to 3.1.11 and afterwards this error seems to be new. 
- When changing the type of rewrite from `forward` to `redirect` it works without any exception.
- I have created a completly empty new crafter 3.1.11 instance to check that this error is not caused by some other code we wrote.

**Comment by jdrossl** [comment]
hi @philipp-brock-thalia, I was not able to reproduce the error using regular pages, could you please provide more information:
- is the target URL external to Crafter?
- do you have any additional filter that could be affecting?

**Comment by philipp-brock-thalia** [comment]
Hi @jdrossl,
I have just started a blank Crafter 3.1.11 from git by checking out the repo and building it with gradlew. Then I have chosen an empty project and added two sides `test` and `test-new`. Afterwards I configured the mentioned rewrite with forwarding in `/config/engine/urlrewrite.xml`:

    <urlrewrite>
        <rule>
            <from>^/test$</from>
            <to type="forward">/test-new</to>
        </rule>
    </urlrewrite>

Then I was already able to see the Exception in the `localhost.2021-01-XX.log` everytime I have loaded the page.

So to answer your questions:

- No the target is also internal
- There are no additional filters (therefore I have tried in a newly created project)
